### PR TITLE
Register device for tesla telemetry

### PR DESCRIPTION
**Pull request by 0xdev22**
# Proposed Changes

- Implementation of tesla fleet functionality to register a vehicle for open telemetry
- Endpoint to allow different integrations register device for Telemetry
- Unit tests for tesla fleet open telemetry
- Unit test for endpoint to enable telemetry/subscribe command

### Impacted Routes
`/user/devices/:userDeviceID/integrations/:integrationID/commands/telemetry/subscribe`

### CODE REVIEW

**Comment by elffjs** [comment]
Is there a way for the frontend to tell whether the config has been sent/accepted?

**Comment by 0xdev22** [comment]
> Is there a way for the frontend to tell whether the config has been sent/accepted? How do they know whether to call this endpoint, I suppose is the question.

The status endpoint from #311 provides this information. If the virtual token is connected and the endpoint confirms it, they know to call this endpoint

**Comment by elffjs** [comment]
> > Is there a way for the frontend to tell whether the config has been sent/accepted? How do they know whether to call this endpoint, I suppose is the question.
> 
> The status endpoint from #311 provides this information. If the virtual token is connected and the endpoint confirms it, they know to call this endpoint

How do they know that they didn't already call it?

File: internal/services/tesla_fleet_api_service.go [hunk 1]
```diff
@@ -247,6 +269,56 @@ func (t *teslaFleetAPIService) RefreshToken(ctx context.Context, refreshToken st
...
+}
+
+func (t *teslaFleetAPIService) SubscribeForTelemetryData(ctx context.Context, token, region, vin string) error {
+	baseURL := fmt.Sprintf(t.Settings.TeslaFleetURL, region)
+	u, err := url.JoinPath(baseURL, "/api/1/vehicles/fleet_telemetry_config")
+	if err != nil {
+		return err
+	}
+
+	exp := time.Now().AddDate(0, 0, 364).Unix()
```
Comment (elffjs) [inline internal/services/tesla_fleet_api_service.go:293]: From [the docs](https://developer.tesla.com/docs/fleet-api#fleet_telemetry_config-create)

> If not specified, config will be valid indefinitely.

Maybe we do this?

We will need a way for them to delete later, I suppose.

File: internal/services/tesla_fleet_api_service.go [hunk 2]
```diff
@@ -247,6 +269,56 @@ func (t *teslaFleetAPIService) RefreshToken(ctx context.Context, refreshToken st
...
+	}
+
+	resp, err := t.performRequest(ctx, u, token, http.MethodPost, bytes.NewReader(b))
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
+
+	return nil
```
Comment (elffjs) [inline internal/services/tesla_fleet_api_service.go:319]: Mustn't we check the response? See their sample:
```
{
 "response": {
  "updated_vehicles": 1,
  "skipped_vehicles": {
   "missing_key": [],
   "unsupported_hardware": [
    "5YJ3000000NEXUS02"
   ],
   "unsupported_firmware": [
    "5YJ3000000NEXUS02"
   ]
  }
 }
}
```

File: internal/services/tesla_fleet_api_service_test.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	"time"
+
+	"github.com/jarcoal/httpmock"
+	"github.com/stretchr/testify/suite"
+
+	"github.com/DIMO-Network/devices-api/internal/config"
+	"github.com/DIMO-Network/devices-api/internal/test"
+)
+
+const mockTeslaFleetBaeURL = "https://fleet-mock-api.%s.tesla.com"
```
Comment (elffjs) [inline internal/services/tesla_fleet_api_service_test.go:18]: Bae.

File: internal/utils/utils.go [hunk 4]
```diff
@@ -31,3 +32,9 @@ func GetSliceDiff(subset, superset []string) []string {
 	}
 	return diffs
 }
+
+func ExistsInSlice(needle string, haystack []string) bool {
```
Comment (elffjs) [inline internal/utils/utils.go:36]: Does this do more than [`slices.Contains`](https://pkg.go.dev/golang.org/x/exp/slices#Contains)?

File: internal/controllers/user_integrations_controller.go [hunk 5]
```diff
@@ -513,6 +514,110 @@ func (udc *UserDevicesController) OpenFrunk(c *fiber.Ctx) error {
...
+// @Param       userDeviceID  path string true "Device ID"
+// @Param       integrationID path string true "Integration ID"
+// @Router      /user/devices/{userDeviceID}/integrations/{integrationID}/commands/telemetry/subscribe [post]
+func (udc *UserDevicesController) TelemetrySubscribe(c *fiber.Ctx) error {
+	userDeviceID := c.Params("userDeviceID")
+	integrationID := c.Params("integrationID")
+
+	logger := helpers.GetLogger(c, udc.log).With().
+		Str("IntegrationID", integrationID).
+		Str("UserDeviceID", userDeviceID).
```
Comment (elffjs) [inline internal/controllers/user_integrations_controller.go:533]: I believe `userDeviceId` will already be in there.

File: internal/controllers/user_integrations_controller.go [hunk 6]
```diff
@@ -513,6 +514,110 @@ func (udc *UserDevicesController) OpenFrunk(c *fiber.Ctx) error {
...
+		return opaqueInternalError
+	}
+
+	if md.TeslaRegion == "" || md.Commands == nil {
+		return fiber.NewError(fiber.StatusBadRequest, "No commands config for integration and device")
+	}
+
+	if len(md.Commands.Capable) == 0 {
+		return fiber.NewError(fiber.StatusBadRequest, "Telemetry command not available for device and integration combination")
+	}
```
Comment (elffjs) [inline internal/controllers/user_integrations_controller.go:578]: > Telemetry command not available for device and integration combination

Is this what we've checked? I would expect some "does this slice contain `telemetry/subscribe`".

File: internal/controllers/user_integrations_controller.go [hunk 7]
```diff
@@ -513,6 +514,110 @@ func (udc *UserDevicesController) OpenFrunk(c *fiber.Ctx) error {
...
+		udai.Metadata = null.JSONFrom(newMeta)
+		_, err = udai.Update(c.Context(), udc.DBS().Writer, boil.Whitelist(models.UserDeviceAPIIntegrationColumns.Metadata))
+		if err != nil {
+			return fiber.NewError(fiber.StatusInternalServerError, "could not save command state", err.Error())
+		}
+	default:
+		return fiber.NewError(fiber.StatusBadRequest, "Integration not supported for this command")
+	}
+
+	logger.Info().Msg("Successfully enqueued command.")
```
Comment (elffjs) [inline internal/controllers/user_integrations_controller.go:616]: Maybe a different log here? "Successfully subscribed to telemetry" or something.

File: internal/services/tesla_fleet_api_service.go (hunk 2 above)
Comment (elffjs) [reply internal/services/tesla_fleet_api_service.go:319]: I guess in this case we can look for `updated_vehicles` being 1, and if not we can see what list it's on.

File: internal/controllers/user_integrations_controller.go [hunk 8]
```diff
@@ -513,6 +514,110 @@ func (udc *UserDevicesController) OpenFrunk(c *fiber.Ctx) error {
...
+
+	switch integration.Vendor {
+	case constants.TeslaVendor:
+		if err := udc.teslaFleetAPISvc.SubscribeForTelemetryData(c.Context(),
+			udai.AccessToken.String,
+			md.TeslaRegion,
+			device.VinIdentifier.String,
+		); err != nil {
+			logger.Error().Err(err).Msg("error registering for telemetry")
+			return fiber.NewError(fiber.StatusFailedDependency, "could not register device for tesla telemetry: ", err.Error())
```
Comment (elffjs) [inline internal/controllers/user_integrations_controller.go:598]: I've never been big on these WebDAV status codes, but let's see how it goes.

**Pull request merged**